### PR TITLE
[SQL] SPARK-6489: Optimize lateral view with explode to not unnecessary columns.

### PR DESCRIPTION
**Pull request by dreamquster**

### CODE REVIEW

**Comment by dreamquster** [comment]
What's wrong?Who can verify it?

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -171,8 +171,23 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(substitutedProjection, child)
 
+    case gen@Generate(generator: Explode, isJoin, isOuter, alias, child) =>
+      val allReferences = gen.references ++ gen.parentReferences
+      val pruneProject = prunedChild(child, allReferences)
+      Generate(generator, isJoin, isOuter, alias, pruneProject)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Is that OK if we just do like
```scala
case gen@Generate(_, _, _, _, child) 
  if child.outputSet -- allReferences.filter(c.outputSet.contains)).nonEmpty =>
  gen.copy(child = Project(a.references.toSeq, child))
```

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Then we don't need other code change except the unit test.

Comment (dreamquster) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Ok,thanks:)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 2]
```diff
@@ -55,6 +55,8 @@ case class Generate(
     child: LogicalPlan)
   extends UnaryNode {
 
+  var parentReferences:AttributeSet = AttributeSet(Seq.empty)
+
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: This mutable state is no longer needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -171,8 +171,26 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
+
+    case plan =>
+      plan.children.foreach { c =>
+        c match {
+          case gen@Generate(generator: Explode, isJoin, isOuter, alias, child) =>
+            gen.parentReferences = plan.references;
+          case _ => // nothing
+        }
+      }
+      plan
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This change is no longer needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -171,8 +171,26 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(substitutedProjection, child)
 
+    case gen@Generate(_: Explode, _, _, _, c) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Is there any reason to only match on `Explode`?

Comment (dreamquster) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: SPARK-6489:this issue means it pruned columns when explode lateral view clause.
Does it need to extend to all generate clause? I want to mark generate node's parent reference in analysis phase. Is there another gracefully method to get  its parent references?

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Oh, I see.  Sorry I did not look closely enough.  In general we do this by matching on a project that is above the node in question.  This works because projects will get pushed down or created in case where unneeded attributes exist.  There are other examples in this file where we do this.

**Comment by dreamquster** [comment]
What's wrong with  the code  in PruningSuite?

````scala
createPruningTest("Column pruning - explode with aggregate",
    "SELECT name, sum(d) AS sumd FROM person LATERAL VIEW explode(data) d AS d GROUP BY name",
    Seq("name", "sumd"),
    Seq("name","data"),
    Seq.empty)

  createPruningTest("Column pruning - outer explode with limit",
    "SELECT name FROM person LATERAL VIEW OUTER explode(data) outd AS d" +
      " where  name < \"C\" limit 3",
    Seq("name"),
    Seq("name", "data"),
    Seq.empty)

  createPruningTest(s"Column pruning - select all without explode optimze - query test",
    "SELECT * FROM person LATERAL VIEW OUTER explode(data) outd AS d WHERE 20 < age",
    Seq("name", "age", "data", "d"),
    Seq("name", "age", "data"),
    Seq.empty)
`````

jenkins report -> Error: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.mr.MapRedTask

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -171,6 +172,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(substitutedProjection, child)
 
+    // add a project which is blocked by Generate
+    case p @ pushBelowGenerate(newChild) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Just look for Project on top of generate instead of attempting to walk the query tree.  You can assume other rules will push projects down to you.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala [hunk 6]
```diff
@@ -502,7 +502,11 @@ class FilterPushdownSuite extends PlanTest {
...
     }
     val optimized = Optimize(originalQuery)
-
-    comparePlans(optimized, originalQuery)
+    val correctAnswer = {
+      testRelationWithArrayType
+        .generate(Explode(Seq("c"), 'c_arr), true, false, Some("arr"))
+        .where(('c > 6) || ('b > 5)).analyze
+    }
+    comparePlans(optimized, correctAnswer)
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala]: What is going on here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala [hunk 7]
```diff
@@ -56,7 +56,8 @@ case class Generate(
   val boundGenerator = BindReferences.bindReference(generator, child.output)
 
   override def execute(): RDD[Row] = {
-    if (join) {
+    // #SPARK-6489 do not join when the child has no output
+    if (join && child.output.nonEmpty) {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala]: This seems unrelated.  Also, how is it possible to run a generator when there is no output?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala [hunk 8]
```diff
@@ -87,6 +87,27 @@ class PruningSuite extends HiveComparisonTest with BeforeAndAfter {
     Seq("key"),
     Seq.empty)
 
+  createPruningTest("Column pruning - explode with aggregate",
+    "SELECT name, sum(d) AS sumd FROM person LATERAL VIEW explode(data) d AS d GROUP BY name",
+    Seq("name", "sumd"),
+    Seq("name","data"),
+    Seq.empty)
```
Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: I don't think we need to go all the way to hive to create an integration test.  Instead create some unit tests in `...catalyst.optimizer.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 5 above)
Comment (dreamquster) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Generate block the process of collectProjectsAndFilters in PhysicalOperation.
Except that Project on top of generate, the expression named Filter may also be on the top of generate. How to solve this scene?
Project
    Filter
      Generate(explode)
Are there  many Filters between Project and Generate?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala (hunk 7 above)
Comment (dreamquster) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala]: SELECT one FROM person LATERAL VIEW explode(1)  AS one;
Project do not need the table data,  and know the table count of rows.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 5 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It is possible that we need several rules to accomplish all of the various optimizations.  However, as its written now this is trying to do too much and as a result is too hard to follow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala (hunk 7 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala]: We should not be trying to optimize degenerate queries such as this one by putting random hacks into the execution engine.

**Comment by rxin** [comment]
@cloud-fan can you take this over?

**Comment by andrewor14** [comment]
Looks like #8286 already supersedes this. I think we can safely close this patch.

**Comment by rxin** [comment]
I'm going to close this pull request.
